File: chazm-api/src/main/java/runtimemodels/chazm/api/entity/manager/InstanceGoalManager.java
package runtimemodels.chazm.api.entity.manager;

import runtimemodels.chazm.api.entity.InstanceGoal;
import runtimemodels.chazm.api.id.UniqueId;

import java.util.Collection;
import java.util.Set;

/**
 * The {@linkplain InstanceGoalManager} interface defines the necessary APIs for managing {@linkplain InstanceGoal}s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public interface InstanceGoalManager {
    /**
     * Adds an {@linkplain InstanceGoal} to this {@linkplain InstanceGoalManager}.
     *
     * @param goal the {@linkplain InstanceGoal} to add.
     */
    void addInstanceGoal(InstanceGoal goal);

    /**
     * Adds a set of {@linkplain InstanceGoal}s to this {@linkplain InstanceGoalManager}.
     *
     * @param goals the set of {@linkplain InstanceGoal} to add.
     */
    void addInstanceGoals(Collection<InstanceGoal> goals);

    /**
     * Returns an {@linkplain InstanceGoal} from this {@linkplain InstanceGoalManager}.
     *
     * @param id the {@linkplain UniqueId} that represents the {@linkplain InstanceGoal} to retrieve.
     * @return the {@linkplain InstanceGoal} if it exists, <code>null</code> otherwise.
     */
    InstanceGoal getInstanceGoal(UniqueId<InstanceGoal> id);

    /**
     * Returns a set of {@linkplain InstanceGoal} from this {@linkplain InstanceGoalManager}.
     *
     * @return the set of {@linkplain InstanceGoal}.
     */
    Set<InstanceGoal> getInstanceGoals();

    /**
     * Removes an {@linkplain InstanceGoal} from this {@linkplain InstanceGoalManager}.
     *
     * @param id the {@linkplain UniqueId} that represents the {@linkplain InstanceGoal} to remove.
     */
    void removeInstanceGoal(UniqueId<InstanceGoal> id);

    /**
     * Removes a set of {@linkplain InstanceGoal}s from this {@linkplain InstanceGoalManager}.
     *
     * @param ids the set of {@linkplain UniqueId} that represents the {@linkplain InstanceGoal}s to remove.
     */
    void removeInstanceGoals(Collection<UniqueId<InstanceGoal>> ids);

    /**
     * Removes all {@linkplain InstanceGoal}s from this {@linkplain InstanceGoalManager}.
     */
    void removeAllInstanceGoals();
}

File: chazm-api/src/main/kotlin/runtimemodels/chazm/api/entity/InstanceGoalManager.kt
package runtimemodels.chazm.api.entity

import runtimemodels.chazm.api.id.InstanceGoalId
import runtimemodels.chazm.api.id.SpecificationGoalId

/**
 * The [InstanceGoalManager] interface defines APIs that can be used for managing a set [InstanceGoal]s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
interface InstanceGoalManager : Map<InstanceGoalId, InstanceGoal> {
    /**
     * Adds an [InstanceGoal] to this [InstanceGoalManager].
     *
     * @param goal the [InstanceGoal] to add.
     */
    fun add(goal: InstanceGoal)

    /**
     * Removes an [InstanceGoal] from this [InstanceGoalManager].
     *
     * @param id the [InstanceGoalId] that represents the [InstanceGoal] to remove.
     * @return the [InstanceGoal] that was removed.
     */
    fun remove(id: InstanceGoalId): InstanceGoal

    /**
     * Removes all [InstanceGoal]s associated to a [SpecificationGoal] from this [InstanceGoalManager].
     *
     * @param id the [SpecificationGoalId] that represents the [SpecificationGoal]
     */
    fun remove(id: SpecificationGoalId)
}
